File: frontend/src/types/project.ts
export interface ProjectOut {
  id: number
  project_name: string
  description?: string
  language?: string
  status: string
  file_count: number
  source_mode?: 'files' | 'audit_archive'
  source_malware_status?: 'clean' | 'infected' | 'degraded' | 'error' | null
  can_update: boolean
  can_delete: boolean
  last_review_at?: string
  /** v2.0: 最近一次成功审查的真实评分,无审查记录时为 null */
  score?: number | null
  create_time: string
}

export interface ProjectSourceArchiveOut {
  original_filename: string
  archive_sha256: string
  compressed_size: number
  expanded_size: number
  file_count: number
  max_member_size: number
  max_compression_ratio: number
  storage_status: string
  malware_status: 'clean' | 'infected' | 'degraded' | 'error'
  audit_status: 'not_started' | 'queued' | 'running' | 'succeeded' | 'failed' | 'blocked' | 'cancelled'
  audit_started_at?: string | null
  audit_heartbeat_at?: string | null
  audit_completed_at?: string | null
  quarantined: boolean
  threat_count: number
}

export interface ProjectDetailOut {
  id: number
  project_name: string
  description?: string
  language?: string
  status: string
  file_count: number
  source_mode: 'files' | 'audit_archive'
  source_archive?: ProjectSourceArchiveOut | null
  can_update: boolean
  can_delete: boolean
  create_time: string
  update_time: string
  recent_tasks: { id: number; score: number; total_issues: number; status: string; create_time: string }[]
}

export interface CodeFileOut {
  id: number
  project_id: number
  file_name: string
  file_path?: string
  language: string
  size_bytes: number
  line_count: number
  version_no: number
  /** v2: 是否二进制文件(0否 1是),二进制文件不展示编辑器,改用下载接口 */
  is_binary: number
  /** v3: 文件 MIME 类型(如 image/png、application/zip),用于前端图标识别 */
  mime_type?: string
  /** v3: 原始字节数(对二进制文件等于 original_blob 长度,文本文件等于 content 编码长度) */
  raw_size?: number
  create_time: string
  update_time: string
}

export interface CodeFileDetailOut extends CodeFileOut {
  content: string
  /** v3: 文件 MD5 摘要(后端 T06 入库时计算) */
  md5_hash?: string
  /** v3: 文件 SHA-256 摘要 */
  sha256_hash?: string
}

/**
 * 文件元信息(用于二进制文件展示)
 * 不含内容,仅元数据
 */
export interface CodeFileMetaOut {
  id: number
  file_name: string
  file_path?: string
  language: string
  size_bytes: number
  raw_size?: number
  line_count: number
  version_no: number
  is_binary: number
  mime_type?: string
  md5_hash?: string
  sha256_hash?: string
  create_time: string
  update_time: string
}

export interface VersionOut {
  version_no: number
  change_desc?: string
  operator_id?: number
  create_time: string
}

export interface VersionDetailOut {
  file_id: number
  version_no: number
  content: string
  change_desc?: string
  create_time: string
}

export interface ProjectCreateIn {
  project_name: string
  description?: string
  language?: string
}

export interface ProjectUpdateIn {
  project_name?: string
  description?: string
  language?: string
  status?: string
}

export interface CodeFileCreateIn {
  file_name: string
  file_path?: string
  content: string
  language?: string
}

export interface CodeFileUpdateIn {
  content: string
  change_desc?: string
}

export interface CodeFileRenameIn {
  file_name: string
}

export interface CodeFileUploadOut {
  file_name: string
  size_bytes: number
  language: string
}

File: backend/app/schemas/project.py
"""
项目模块Pydantic Schema
"""
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class ProjectIn(BaseModel):
    """创建项目请求体"""
    project_name: str = Field(min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=500)
    language: Optional[str] = Field(default=None, max_length=50)


class ProjectUpdateIn(BaseModel):
    """更新项目请求体"""
    project_name: Optional[str] = Field(default=None, min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=500)
    language: Optional[str] = Field(default=None, max_length=50)
    status: Optional[str] = Field(default=None, pattern="^(active|archived)$")


class RemoteProjectImportIn(BaseModel):
    """公开 HTTPS 源码归档导入请求。"""

    url: str = Field(min_length=12, max_length=2048)
    project_name: str = Field(min_length=1, max_length=100)
    description: Optional[str] = Field(default=None, max_length=500)
    language: Optional[str] = Field(default=None, max_length=50)
    audit_mode: bool = False


class RemoteProjectImportTaskOut(BaseModel):
    """可恢复远程导入任务的公开状态。"""

    task_id: str
    status: Literal["queued", "running", "succeeded", "failed"]
    attempt_count: int
    max_attempts: int
    project_id: Optional[int] = None
    result: dict[str, Any] = Field(default_factory=dict)
    error: Optional[dict[str, str]] = None
    next_attempt_at: Optional[datetime] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    create_time: datetime
    update_time: datetime


class ProjectSourceArchiveOut(BaseModel):
    """隔离整包源码的安全摘要，不包含原包字节或源码。"""

    original_filename: str
    archive_sha256: str
    compressed_size: int
    expanded_size: int
    file_count: int
    max_member_size: int
    max_compression_ratio: float
    storage_status: str
    malware_status: str
    audit_status: str
    audit_started_at: Optional[datetime] = None
    audit_heartbeat_at: Optional[datetime] = None
    audit_completed_at: Optional[datetime] = None
    quarantined: bool
    threat_count: int


class ProjectOut(BaseModel):
    """项目列表项响应

    v2.0 新增 `score`: 最近一次成功审查的评分,前端 ProjectList 使用真实评分,
    不再 hash 派生假数字。无审查记录时为 None。
    R6 修复(2026-06-25):补齐 update_time,对齐 Project ORM。
    """
    id: int
    project_name: str
    description: Optional[str] = None
    language: Optional[str] = None
    status: str
    file_count: int = 0
    source_mode: str = "files"
    source_malware_status: Optional[str] = None
    can_update: bool = False
    can_delete: bool = False
    last_review_at: Optional[datetime] = None
    score: Optional[int] = None
    agent_run_count: int = 0
    last_agent_run_at: Optional[datetime] = None
    create_time: datetime
    # R6 修复:补齐 update_time,对齐 Project ORM
    update_time: Optional[datetime] = None

    model_config = {"from_attributes": True}


class RecentTaskOut(BaseModel):
    """最近审查任务概要"""
    id: int
    score: int
    total_issues: int
    status: str
    create_time: datetime


class SourceRevisionOut(BaseModel):
    """项目源码修复副本摘要。"""
    id: int
    revision_no: int
    source_sha256: str
    parent_sha256: Optional[str] = None
    repaired_files: list[str] = Field(default_factory=list)
    repair_notes: str = ""
    create_time: Optional[datetime] = None


class ProjectDetailOut(BaseModel):
    """项目详情响应"""
    id: int
    project_name: str
    description: Optional[str] = None
    language: Optional[str] = None
    status: str
    file_count: int = 0
    source_mode: str = "files"
    source_archive: Optional[ProjectSourceArchiveOut] = None
    can_update: bool = False
    can_delete: bool = False
    agent_run_count: int = 0
    last_agent_run_at: Optional[datetime] = None
    create_time: datetime
    update_time: datetime
    recent_tasks: list[RecentTaskOut] = []
    source_revisions: list[SourceRevisionOut] = Field(default_factory=list)

    model_config = {"from_attributes": True}
